File: services/client/src/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"fmt"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/lottery"
)

const (
	TYPE_BET byte = 1
	TYPE_END byte = 2

	HEADER_LEN     = 4
	MIN_PACKET_LEN = 20
)

func is_valid_type(t byte) bool {
	return t == TYPE_BET || t == TYPE_END
}

func ParseMessageType(packet []byte) (byte, error) {
	messageType := packet[0]
	if !is_valid_type(messageType) {
		return 0, fmt.Errorf("invalid message type: %d", packet[0])
	}

	return messageType, nil
}

func SerializeEnd(agencyID uint8) []byte {
	packet := make([]byte, 0, 4)
	packet = binary.BigEndian.AppendUint16(packet, 2)
	packet = append(packet, TYPE_END, agencyID)
	return packet
}

func IsEndPacket(data []byte) bool {
	return len(data) == 1 && data[0] == TYPE_END
}

func SerializeBet(bet lottery.Bet) []byte {
	betData := make([]byte, 0, 100)

	firstNameLen := len(bet.FirstName)
	lastNameLen := len(bet.LastName)

	payloadLen := MIN_PACKET_LEN + firstNameLen + lastNameLen
	betData = binary.BigEndian.AppendUint16(betData, uint16(payloadLen)) // 2B

	betData = append(betData, TYPE_BET)            // 1B
	betData = append(betData, bet.AgencyId)        // 1B
	betData = append(betData, uint8(firstNameLen)) // 1B
	betData = append(betData, uint8(lastNameLen))  // 1B

	betData = append(betData, []byte(bet.FirstName)...)            // Variable length
	betData = append(betData, []byte(bet.LastName)...)             // Variable length
	betData = append(betData, []byte(bet.Birthdate)...)            // 10B
	betData = binary.BigEndian.AppendUint32(betData, bet.Document) // 4B
	betData = binary.BigEndian.AppendUint16(betData, bet.Number)   // 2B

	return betData
}

func DeserializeBet(betData []byte) (lottery.Bet, error) {
	if len(betData) < HEADER_LEN {
		return lottery.Bet{}, fmt.Errorf("packet too short")
	}

	messageType, err := ParseMessageType(betData)
	if err != nil {
		return lottery.Bet{}, err
	}

	if messageType != TYPE_BET {
		return lottery.Bet{}, fmt.Errorf("invalid bet type: %d", betData[0])
	}

	agencyId := betData[1]
	firstNameLen := int(betData[2])
	lastNameLen := int(betData[3])

	if len(betData) != MIN_PACKET_LEN+firstNameLen+lastNameLen {
		return lottery.Bet{}, fmt.Errorf("invalid bet data: wrong size")
	}

	const firstNameStart = 4
	const birthdateLen = 10
	const documentLen = 4

	lastNameStart := firstNameStart + firstNameLen
	birthdateStart := lastNameStart + lastNameLen
	documentStart := birthdateStart + birthdateLen
	numberStart := documentStart + documentLen

	firstName := string(betData[firstNameStart:lastNameStart])
	lastName := string(betData[lastNameStart:birthdateStart])
	birthdate := string(betData[birthdateStart:documentStart])
	document := binary.BigEndian.Uint32(betData[documentStart:numberStart])
	number := binary.BigEndian.Uint16(betData[numberStart:])

	return lottery.Bet{
		AgencyId:  agencyId,
		FirstName: firstName,
		LastName:  lastName,
		Document:  document,
		Birthdate: birthdate,
		Number:    number,
	}, nil
}

File: services/server/src/protocol/protocol.py
from lottery import Bet

TYPE_BET = 1
TYPE_END = 2

HEADER_LEN = 4
MIN_PACKET_LEN = 20

def parse_message_type(packet: bytes) -> int:
    message_type = packet[0]
    if message_type not in (TYPE_BET, TYPE_END):
        raise ValueError(f"invalid message type: {message_type}")

    return message_type


def serialize_end() -> bytes:
    payload = bytearray([TYPE_END])
    return len(payload).to_bytes(2, byteorder="big") + bytes(payload)


def is_end_packet(packet: bytes) -> bool:
    return len(packet) == 2 and packet[0] == TYPE_END


def serialize_bet(bet) -> bytes:
    first_name = bet.first_name.encode("utf-8")
    last_name = bet.last_name.encode("utf-8")
    birthdate = bet.birthdate.encode("utf-8")

    first_name_len = len(first_name)
    last_name_len = len(last_name)

    payload_len = MIN_PACKET_LEN + first_name_len + last_name_len

    bet_data = bytearray()

    bet_data.extend(payload_len.to_bytes(2, byteorder="big"))

    bet_data.append(TYPE_BET)
    bet_data.append(bet.agency_id)
    bet_data.append(first_name_len)
    bet_data.append(last_name_len)

    bet_data.extend(first_name)
    bet_data.extend(last_name)
    bet_data.extend(birthdate)
    bet_data.extend(bet.document.to_bytes(4, byteorder="big"))
    bet_data.extend(bet.number.to_bytes(2, byteorder="big"))

    return bytes(bet_data)


def deserialize_bet(bet_data: bytes):
    if len(bet_data) < HEADER_LEN:
        raise ValueError("packet too short")

    message_type = parse_message_type(bet_data)
    if message_type != TYPE_BET:
        raise ValueError(f"invalid bet type: {bet_data[0]}")

    agency_id = bet_data[1]
    first_name_len = bet_data[2]
    last_name_len = bet_data[3]

    if len(bet_data) != MIN_PACKET_LEN + first_name_len + last_name_len:
        raise ValueError("invalid bet data: wrong size")

    first_name_start = 4
    birthdate_len = 10
    document_len = 4

    last_name_start = first_name_start + first_name_len
    birthdate_start = last_name_start + last_name_len
    document_start = birthdate_start + birthdate_len
    number_start = document_start + document_len

    first_name = bet_data[first_name_start:last_name_start].decode("utf-8")
    last_name = bet_data[last_name_start:birthdate_start].decode("utf-8")
    birthdate = bet_data[birthdate_start:document_start].decode("utf-8")

    document = int.from_bytes(
        bet_data[document_start:number_start],
        byteorder="big"
    )

    number = int.from_bytes(
        bet_data[number_start:],
        byteorder="big"
    )

    return Bet(
        agency_id=agency_id,
        first_name=first_name,
        last_name=last_name,
        document=document,
        birthdate=birthdate,
        number=number,
    )
